WebKit Should Emit Signposts During Scrolling Performance Testing

connorpobrien opened the pull request:
#### 30a58f4978a9ccd95b115ae8c890fe43f2fa48f5
<pre>
WebKit Should Emit Signposts During Scrolling Performance Testing
<a href="https://bugs.webkit.org/show_bug.cgi?id=277894">https://bugs.webkit.org/show_bug.cgi?id=277894</a>
<a href="https://rdar.apple.com/133582534">rdar://133582534</a>

Reviewed by Simon Fraser.

This change adds signpost logging during scrolling performance testing.

* Source/WTF/wtf/SystemTracing.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
(WebKit::RemoteLayerTreeEventDispatcher::handleWheelEvent):
(WebKit::RemoteLayerTreeEventDispatcher::startFingerDownSignpostInterval):
(WebKit::RemoteLayerTreeEventDispatcher::endFingerDownSignpostInterval):
(WebKit::RemoteLayerTreeEventDispatcher::startMomentumSignpostInterval):
(WebKit::RemoteLayerTreeEventDispatcher::endMomentumSignpostInterval):
(WebKit::RemoteLayerTreeEventDispatcher::handleSyntheticWheelEvent):
(WebKit::RemoteLayerTreeEventDispatcher::didStartRubberbanding):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::setRubberBandingInProgressForNode):

Canonical link: <a href="https://commits.webkit.org/282437@main">https://commits.webkit.org/282437@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9529e03929211c9b938e278e476b88b166b7e67d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63162 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42518 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15758 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67183 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13770 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50205 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14050 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50910 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/9513 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66231 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39486 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54693 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31588 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36171 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/12040 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12642 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/56276 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57711 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12369 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68878 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/62405 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7108 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11989 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/58222 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7140 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54765 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/58431 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13996 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5933 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/84168 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38338 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14830 "Found 1 new JSC stress test failure: stress/sampling-profiler-display-name.js.dfg-eager-no-cjit-validate (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39418 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40529 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39160 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

GeorgesStavracas commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
```diff
@@ -189,6 +189,12 @@ void RemoteLayerTreeEventDispatcher::handleWheelEvent(const WebWheelEvent& wheel
+    auto scrollingTree = this->scrollingTree();
+    if (scrollingTree->scrollingPerformanceTestingEnabled()) {
+        if (wheelEvent.phase() == WebWheelEvent::PhaseBegan)
+        WTFBeginSignpostAlways(nullptr, ScrollingPerformanceTestScrollInterval, "isAnimation=YES;");
```
Alignment seems off here

connorpobrien replied:
Thank you

smfr replied:
Still wrong; it needs to be indented.

smfr replied:
I'm not convinced that you call  `WTFEndSignpostAlways(..., ScrollingPerformanceTestScrollInterval, ...)` for gestures that don't involve rubberbanding or momentum.

connorpobrien replied:
Thanks. Working now on confirming if the interval ends without rubber banding or momentum

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
```diff
@@ -651,6 +657,21 @@ void RemoteLayerTreeEventDispatcher::handleSyntheticWheelEvent(PageIdentifier pa
+        if (!m_ScrollingPerformanceTestSignpostEndIntervals && !std::abs(event.delta().height())) {
+            m_ScrollingPerformanceTestSignpostEndIntervals++;
+            WTFEndSignpostAlways(nullptr, ScrollingPerformanceTestScrollInterval, "isAnimation=YES;");
+            WTFEndSignpostAlways(nullptr, ScrollingPerformanceTestMomentumInterval, "isAnimation=Yes;");
```
Why is the capitalization of YES different on these two lines?

connorpobrien replied:
Typo. Fixing now

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
```diff
@@ -651,6 +657,21 @@ void RemoteLayerTreeEventDispatcher::handleSyntheticWheelEvent(PageIdentifier pa
+        }
+
+        if (m_ScrollingPerformanceTestSignpostEndIntervals > 0 && event.momentumPhase() == WebWheelEvent::PhaseEnded)
+            m_ScrollingPerformanceTestSignpostEndIntervals = 0;
```
Do we clear this correctly if we hit rubberbanding outside the momentum phase?

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
```diff
@@ -659,6 +680,15 @@ void RemoteLayerTreeEventDispatcher::handleSyntheticWheelEvent(PageIdentifier pa
+{
+    if (!m_ScrollingPerformanceTestSignpostEndIntervals) {
+        WTFEndSignpostAlways(nullptr, ScrollingPerformanceTestScrollInterval, "isAnimation=YES;");
+        WTFEndSignpostAlways(nullptr, ScrollingPerformanceTestMomentumInterval, "isAnimation=Yes;");
```
Inconsistent capitalization.

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
```diff
@@ -135,6 +135,8 @@ class RemoteLayerTreeEventDispatcher : public ThreadSafeRefCounted<RemoteLayerTr
 
     void waitForRenderingUpdateCompletionOrTimeout() WTF_REQUIRES_LOCK(m_scrollingTreeLock);
 
+    void scrollingPerformanceTestLogRubberBanding();
```
Let's just call this `didStartRubberbanding()`

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
```diff
@@ -159,6 +161,8 @@ class RemoteLayerTreeEventDispatcher : public ThreadSafeRefCounted<RemoteLayerTr
     std::optional<DisplayLinkObserverID> m_displayRefreshObserverID;
     PAL::HysteresisActivity m_wheelEventActivityHysteresis;
 
+    int m_ScrollingPerformanceTestSignpostEndIntervals = 0; // Used to prevent multiple end signpost calls
```
I think this is accessed from different threads, so needs to be `std::atomic<int>`

The name is rather unclear. What does it really represent? Is it counting the end of `ScrollingPerformanceTestScrollInterval` or `ScrollingPerformanceTestMomentumInterval`?

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
```diff
@@ -159,6 +161,9 @@ class RemoteLayerTreeEventDispatcher : public ThreadSafeRefCounted<RemoteLayerTr
     std::optional<DisplayLinkObserverID> m_displayRefreshObserverID;
     PAL::HysteresisActivity m_wheelEventActivityHysteresis;
 
+    std::atomic<bool> m_FingerDownIntervalIsActive = false;
```
`m_fingerDownIntervalIsActive` etc (lowercase the letter after the underscore).

connorpobrien replied:
Fixed thanks

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
```diff
@@ -659,6 +683,19 @@ void RemoteLayerTreeEventDispatcher::handleSyntheticWheelEvent(PageIdentifier pa
+void RemoteLayerTreeEventDispatcher::didStartRubberbanding()
+{
+    if (m_MomentumIntervalIsActive) {
+        WTFEndSignpostAlways(nullptr, ScrollingPerformanceTestMomentumInterval, "isAnimation=YES;");
```
These two lines repeat in two places, making things error-prone to future edits. Consider a helper function for them and the FingerDown version.

connorpobrien replied:
I added 4 helper functions for start/end fingerDown/momentum

smfr commented on Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
```diff
@@ -28,6 +28,7 @@
 
 #if PLATFORM(MAC) && ENABLE(UI_SIDE_COMPOSITING)
 
+#import "MomentumEventDispatcher.h"
```
I don't think you need either of the new #includes.

connorpobrien replied:
Fixed thanks

connorpobrien committed:
WebKit Should Emit Signposts During Scrolling Performance Testing
https://bugs.webkit.org/show_bug.cgi?id=277894
rdar://133582534

Reviewed by Simon Fraser.

This change adds signpost logging during scrolling performance testing.

* Source/WTF/wtf/SystemTracing.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.cpp:
(WebKit::RemoteLayerTreeEventDispatcher::handleWheelEvent):
(WebKit::RemoteLayerTreeEventDispatcher::startFingerDownSignpostInterval):
(WebKit::RemoteLayerTreeEventDispatcher::endFingerDownSignpostInterval):
(WebKit::RemoteLayerTreeEventDispatcher::startMomentumSignpostInterval):
(WebKit::RemoteLayerTreeEventDispatcher::endMomentumSignpostInterval):
(WebKit::RemoteLayerTreeEventDispatcher::handleSyntheticWheelEvent):
(WebKit::RemoteLayerTreeEventDispatcher::didStartRubberbanding):
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteLayerTreeEventDispatcher.h:
* Source/WebKit/UIProcess/RemoteLayerTree/mac/RemoteScrollingCoordinatorProxyMac.mm:
(WebKit::RemoteScrollingCoordinatorProxyMac::setRubberBandingInProgressForNode):

Canonical link: https://commits.webkit.org/282437@main
